File: src/main/java/com/fwdekker/randomness/integer/IntegerSettingsDialog.java
package com.fwdekker.randomness.integer;

import com.fwdekker.randomness.JavaHelperKt;
import com.fwdekker.randomness.SettingsDialog;
import com.fwdekker.randomness.ui.ButtonGroupKt;
import com.fwdekker.randomness.ui.JIntSpinner;
import com.fwdekker.randomness.ui.JLongSpinner;
import com.fwdekker.randomness.ui.JSpinnerRange;
import com.intellij.openapi.ui.ValidationInfo;
import kotlin.Unit;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.ButtonGroup;
import javax.swing.JComponent;
import javax.swing.JPanel;


/**
 * Dialog for settings of random integer generation.
 *
 * @see IntegerSettings
 * @see IntegerSettingsAction
 */
public final class IntegerSettingsDialog extends SettingsDialog<IntegerSettings> {
    private JPanel contentPane;
    private JSpinnerRange valueRange;
    private JLongSpinner minValue;
    private JLongSpinner maxValue;
    private JIntSpinner base;
    private ButtonGroup groupingSeparatorGroup;


    /**
     * Constructs a new {@code IntegerSettingsDialog} that uses the singleton {@code IntegerSettings} instance.
     */
    /* default */ IntegerSettingsDialog() {
        this(IntegerSettings.Companion.getDefault());
    }

    /**
     * Constructs a new {@code IntegerSettingsDialog} that uses the given {@code IntegerSettings} instance.
     *
     * @param settings the settings to manipulate with this dialog
     */
    /* default */ IntegerSettingsDialog(final @NotNull IntegerSettings settings) {
        super(settings);

        loadSettings();
    }


    @Override
    public JPanel getRootPane() {
        return contentPane;
    }

    /**
     * Initialises custom UI components.
     * <p>
     * This method is called by the scene builder at the start of the constructor.
     */
    private void createUIComponents() {
        minValue = new JLongSpinner();
        maxValue = new JLongSpinner();
        base = new JIntSpinner(IntegerSettings.DECIMAL_BASE, IntegerSettings.MIN_BASE, IntegerSettings.MAX_BASE);
        valueRange = new JSpinnerRange(minValue, maxValue, Long.MAX_VALUE);

        base.addChangeListener(event -> {
            final int value = ((JIntSpinner) event.getSource()).getValue();
            final boolean enabled = value == IntegerSettings.DECIMAL_BASE;

            ButtonGroupKt.forEach(
                groupingSeparatorGroup,
                button -> {
                    button.setEnabled(enabled);
                    return Unit.INSTANCE;
                }
            );
        });
    }


    @Override
    public void loadSettings(final @NotNull IntegerSettings settings) {
        minValue.setValue(settings.getMinValue());
        maxValue.setValue(settings.getMaxValue());
        base.setValue(settings.getBase());
        ButtonGroupKt.setValue(groupingSeparatorGroup, settings.getGroupingSeparator());
    }

    @Override
    public void saveSettings(final @NotNull IntegerSettings settings) {
        settings.setMinValue(minValue.getValue());
        settings.setMaxValue(maxValue.getValue());
        settings.setBase(base.getValue());
        settings.safeSetGroupingSeparator(ButtonGroupKt.getValue(groupingSeparatorGroup));
    }

    @Override
    @Nullable
    public ValidationInfo doValidate() {
        return JavaHelperKt.firstNonNull(
            minValue.validateValue(),
            maxValue.validateValue(),
            base.validateValue(),
            valueRange.validateValue()
        );
    }
}

File: src/main/kotlin/com/fwdekker/randomness/integer/IntegerSettingsDialog.kt
package com.fwdekker.randomness.integer

import com.fwdekker.randomness.SettingsDialog
import com.fwdekker.randomness.ui.ButtonGroupHelper
import com.fwdekker.randomness.ui.JLongSpinner
import com.fwdekker.randomness.ui.JSpinnerRange
import java.util.function.Consumer
import javax.swing.ButtonGroup
import javax.swing.JPanel


/**
 * Dialog for settings of random integer generation.
 */
class IntegerSettingsDialog(settings: IntegerSettings = IntegerSettings.default) :
    SettingsDialog<IntegerSettings>(settings) {
    private lateinit var contentPane: JPanel
    private lateinit var valueRange: JSpinnerRange
    private lateinit var minValue: JLongSpinner
    private lateinit var maxValue: JLongSpinner
    private lateinit var base: JLongSpinner
    private lateinit var groupingSeparatorGroup: ButtonGroup


    init {
        init()
        loadSettings()
    }


    override fun createCenterPanel() = contentPane

    /**
     * Initialises custom UI components.
     *
     * This method is called by the scene builder at the start of the constructor.
     */
    private fun createUIComponents() {
        minValue = JLongSpinner()
        maxValue = JLongSpinner()
        base = JLongSpinner(
            IntegerSettings.DECIMAL_BASE.toLong(),
            IntegerSettings.MIN_BASE.toLong(),
            IntegerSettings.MAX_BASE.toLong()
        )
        valueRange = JSpinnerRange(minValue, maxValue, Long.MAX_VALUE.toDouble())

        base.addChangeListener { event ->
            val value = (event.source as JLongSpinner).value
            val enabled = value == IntegerSettings.DECIMAL_BASE.toLong()
            ButtonGroupHelper.forEach(groupingSeparatorGroup, Consumer { button -> button.isEnabled = enabled })
        }
    }

    override fun doValidate() =
        null
            ?: minValue.validateValue()
            ?: maxValue.validateValue()
            ?: base.validateValue()
            ?: valueRange.validateValue()


    override fun loadSettings(settings: IntegerSettings) {
        minValue.value = settings.minValue
        maxValue.value = settings.maxValue
        base.value = settings.base.toLong()
        ButtonGroupHelper.setValue(groupingSeparatorGroup, settings.groupingSeparator.toString())
    }

    override fun saveSettings(settings: IntegerSettings) {
        settings.minValue = minValue.value
        settings.maxValue = maxValue.value
        settings.base = base.value.toInt()
        settings.setGroupingSeparator(ButtonGroupHelper.getValue(groupingSeparatorGroup)!!)
    }
}
